### PR TITLE
Merge basic blocks where possible when generating LLVM IR.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
The instruction count results aren't a win, but there are hints of goodness in the results for cycles, wall-time, max-rss, and especially binary size. The current version only merges the simplest cases, and there are quite a few more cases that can be handled, so I will continue working on them.

**Comment by nnethercote** [comment]
To summarize:
- MIR uses one definition of BBs, and LLVM IR uses another.
- rustc generates reasonable MIR code.
- rustc does a 1-to-1 translation of MIR BBs to LLVM IR BBs, which is reasonable.
- The resulting LLVM IR looks a bit silly and quite sub-optimal, because of the different BB definition.
- The sub-optimality doesn't end up mattering much in terms of compiler perf.
- The sub-optimality also doesn't matter for the output of opt builds, because LLVM can optimize away the extra jumps and the output ends up the same.
- The sub-optimality matters slightly for the output of debug builds, because it causes binaries to be about 0.5% bigger. It may also make them slightly slower, though I haven't measured that and I suspect the effect would be very small, probably less than 0.5%.

**Comment by nnethercote** [comment]
I'm reopening this for further consideration. Even though it didn't make much difference to compiler perf, which was the original motivation, it might still be worth merging.

Pros:
- Small (mostly <1%) binary size improvements, mostly for debug builds.
- Generated LLVM IR is easier to read, due to not having unnecessary branches.
- Generate binary is easier to debug at the machine code level, due to not having unnecessary branches (cc @Amanieu).

Cons:
- Some extra complexity in codegen.
- It's a little harder to see the MIR-to-LLVM-IR mapping.

File: compiler/rustc_codegen_ssa/src/mir/mod.rs [hunk 1]
```diff
@@ -16,6 +16,18 @@ use rustc_middle::mir::traversal;
...
 
+// Used for tracking the state of generated basic blocks.
+enum CachedLlbb<T> {
+    /// Nothing created yet.
+    None,
+
+    /// Has been created.
+    Some(T),
+
+    /// Nothing created yet, and nothing should be.
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/mir/mod.rs:27]:
```suggestion
    /// Nothing should be created.
```

**Review by bjorn3 — COMMENTED** [review]
How is it ensured that a basic block that will be merged in the future is not codegened at an earlier point? While it doesn't happen every time due to traversing mir in reverse postorder, it can still happen and I believe there are assumptions in cg_ssa about blocks being codegened only once somewhere in the code that avoids alloca where possible.

**Comment by nnethercote** [comment]
> How is it ensured that a basic block that will be merged in the future is not codegened at an earlier point?

I think the reverse postorder traversal guarantees this? In combination with the constraint that merging is only possible for BB0 -> BB1 when BB1 is the only successor of BB0 and BB0 is the only predecessor of BB1.

I just added an assertion to check this -- when the successor is marked as `CachedLllbb::Skip`, the assertion checks that its current state is `CachedLllbb::None`. The full test suite passes with this assertion in place.

**Comment by bjorn3** [comment]
Right, makes sense.

@bors r+

**Comment by nnethercote** [comment]
No idea about that failure. Nothing obviously related to this PR's changes. Let's try again.

@bors retry rollup=never

**Comment by riking** [comment]
Failed test is debuginfo/lexical-scope-in-if-let.rs, test error is `error: line not found in debugger output: [...]z = 0n10` but that line is clearly in the debugger output:

```
0:000>  dv
              z = 0n10
              c = 0n789
```

**Comment by nnethercote** [comment]
I have never seen these debuginfo tests before. But I notice that this test has four `cbd-command`/`cbd-check` groups. The second last one, which passes, looks like this:
```
// cdb-command: g
// cdb-command: dv
// cdb-check:[...]y = true
// cdb-check:[...]b = 0n456
// cdb-check:[...]a = 0n123
// cdb-check:[...]x = 0n42
```
and the `dv` output in the log looks like this:
```
2022-11-15T17:13:30.2292723Z 0:000>  dv
2022-11-15T17:13:30.2292998Z               a = 0n123
2022-11-15T17:13:30.2293303Z               x = 0n42
2022-11-15T17:13:30.2293590Z               b = 0n456
2022-11-15T17:13:30.2293871Z               y = true
```
But the last one, which fails, looks like this:
```
// cdb-command: g
// cdb-command: dv
// cdb-check:[...]z = 0n10
// cdb-check:[...]c = 0n789
// cdb-check:[...]y = true
// cdb-check:[...]b = 0n456
// cdb-check:[...]a = 0n123
// cdb-check:[...]x = 0n42
```
and the `dv` output in the log looks like this:
```
2022-11-15T17:13:30.2297142Z 0:000>  dv
2022-11-15T17:13:30.2297406Z               z = 0n10
2022-11-15T17:13:30.2297692Z               c = 0n789
2022-11-15T17:13:30.2297995Z               a = 0n123
2022-11-15T17:13:30.2298267Z               x = 0n42
2022-11-15T17:13:30.2298547Z               b = 0n456
2022-11-15T17:13:30.2298826Z               y = true
2022-11-15T17:13:30.2299118Z 0:000> 
2022-11-15T17:13:30.2299378Z               z = 0n10
2022-11-15T17:13:30.2299661Z               c = 0n789
2022-11-15T17:13:30.2299966Z               a = 0n123
2022-11-15T17:13:30.2300240Z               x = 0n42
2022-11-15T17:13:30.2300519Z               b = 0n456
2022-11-15T17:13:30.2300816Z               y = true
```
I don't know why the values are repeated, but maybe that's the problem? It's semi-plausible that merging basic blocks might somehow affect the debuginfo in a way that causes this...

Unfortunately I don't have access to a Windows machine which makes this hard to debug.

**Comment by nnethercote** [comment]
@ehuss suggested the problem is the ordering of the debugger output. I have tweaked that and added a temporary commit to do more Windows testing on CI.

**Comment by nnethercote** [comment]
Slightly reordering the debuginfo output fixed the test failure.

@bors r=bjorn3

**Comment by Manishearth** [comment]
@bors p=1

going to close the tree for non-nevers for a while

**Pull request merged**